README.md
What is this?
=============

It's a tool that can be used to control features of the
"Lenovo ThinkPad USB Keyboard with TrackPoint".
The official User Guide can be found [here][1] (PDF).

[1]: http://download.lenovo.com/ibmdl/pub/pc/pccbbs/options_iso/45k1918_ug.pdf "User Guide"

Requirements and Limitations
============================

This tool uses the `hidraw` driver and needs at least *Linux 2.6.39* in order
to work. It can only write settings, not read them, so you have to set
everything at the same time.

On most systems the hidraw device files are readable only by root, so you
might need to run this as root.

Usage
=====

    Usage: tpkbdctl [OPTIONS]
    
      -h, --help               show this
      -V, --version            show version
      -d, --device=?           set device. Autodiscovered if unset
      -v                       increase verbosity
      -s, --sensitivity=?      set sensitivity, range 1-255
      -S, --press-speed=?      set press-speed, range 1-255
      -p, --press-to-select    enable press-to-select
      -R, --press-right        select with right button instead of left
      -D, --dragging           enable dragging
      -r, --release-to-select  enable release-to-select


### Sensitivity
The higher the sensitivity, the less force you need to move the mouse cursor.

### Press to select
If this is enabled, pressing down the trackpoint generates a click.

### Press-speed
This setting determines how fast you have to press to generate a click if
*Press to select* is enabled.

### Dragging
By enabling this, you can not only click by pressing down the trackpoint, but
also drag. This depends on *Press to select* being enabled.

### Press right
This changes the *Press to select* feature to generate right-button clicks
instead of left-button ones.

### Release to select
This inaptly named option (the name was adopted from the Windows driver)
enables double-clicking when using *Press to select*.


Building and Installing
=======================

Debian/Ubuntu
-------------

    fakeroot debian/rules binary
    sudo dpkg -i ../tpkbdctl_*.deb

Others
------

    make
    sudo make install

You need Linux headers installed, minimum version 2.6.39.

Permanent Setup
===============

Debian/Ubuntu
-------------

Just customize `/etc/default/tpkbdctl`.

Others
------

To have your preferred configuration set automatically, a udev rule can be
used. Place the following code in  `/etc/udev/rules.d/10-tpkbdctl.rules`.

    SUBSYSTEM=="hidraw", ATTRS{idVendor}=="17ef", ATTRS{idProduct}=="6009", ACTION=="add", RUN+="/etc/udev/tpkbdctl_runner"

Customize this according to your needs and save it as `/etc/udev/tpkbdctl_runner`.
Don't forget to make it executable.

    #!/bin/sh
    
    /usr/bin/tpkbdctl -d ${DEVNAME} -s 192 # your settings here

See also
========

There's a [kernel patch][2] that provides more functionality than this tool.
It has been merged in Linux 3.6.

[2]: https://github.com/bseibold/linux/tree/tpkbd-unstable (Kernel patch)

License
=======

This tool is licensed under the GNU GPL v2.



debian/changelog
tpkbdctl (0.2.3-1) unstable; urgency=low

  * Bugfix: Manually specifying hidraw device didn't work

 -- unknown <unknown@unknown>  Sun, 13 Jan 2012 12:13:18 +0100

tpkbdctl (0.2.2-1) unstable; urgency=low

  * Push version because README.rst was missing on PyPi

 -- unknown <unknown@unknown>  Sat, 29 Dez 2012 11:51:23 +0100

tpkbdctl (0.2.1-1) unstable; urgency=low

  * Small bugfixes

 -- unknown <unknown@unknown>  Fri, 30 Nov 2012 18:21:01 +0100

tpkbdctl (0.2-1) unstable; urgency=low

  * Initial release

 -- unknown <unknown@unknown>  Thu, 29 Nov 2012 00:02:44 +0100


tpkbdctl.c
/* 
 * (c) 2012 Bernhard Seibold <bernhard.seibold@gmail.com>
 * License: GPLv2
 */
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <getopt.h>
#include <libgen.h>

#include <linux/types.h>
#include <linux/input.h>
#include <linux/hiddev.h>
#include <linux/hidraw.h>

#ifndef VERSION
#define VERSION (unknown)
#endif

#define QUOTE(x) #x
#define STR(x) QUOTE(x)

int verbosity = 0;


int find_device() {
	int i, fd, ret, descsize;
	char name[255];
	struct hidraw_devinfo devinfo;
	for (i=0; i<255; i++) {
		snprintf(name, 255, "/dev/hidraw%d", i);
		fd = open(name, O_RDWR|O_NONBLOCK);
		if (fd == -1) {
			if (errno == EACCES)
				fprintf(stderr, "Permission denied on %s\n", name);
			continue;
		}

		ret = ioctl(fd, HIDIOCGRAWINFO, &devinfo);
		if (ret) {
			close(fd);
			continue;
		}
		if (devinfo.vendor != 0x17ef || devinfo.product != 0x6009) {
			close(fd);
			continue;
		}

		ret = ioctl(fd, HIDIOCGRDESCSIZE, &descsize);
		if (ret || descsize < 100) {
			close(fd);
			continue;
		}

		if (verbosity > 0)
			fprintf(stderr, "Found device at %s\n", name);
		return fd;
	}
	return 0;
}


int main(int argc, char* argv[]) {
	int press_to_select = 0;
	int dragging = 0;
	int release_to_select = 0;
	int press_right = 0;
	int sensitivity = 128;
	int press_speed = 0x38;
	int show_help = 0;
	int show_version = 0;
	int fd, ret;
	char *device = NULL;
	char buf[5];

	int opt, opt_idx;
	static struct option longopts[] = {
		{ "device", required_argument, 0, 0 },
		{ "sensitivity", required_argument, 0, 0},
		{ "press_speed", required_argument, 0, 0},
		{ "help", no_argument, 0, 0},
		{ "press-to-select", no_argument, 0, 0},
		{ "dragging", no_argument, 0, 0},
		{ "release-to-select", no_argument, 0, 0},
		{ "press-right", no_argument, 0, 0},
		{ "version", no_argument, 0, 0},
		{ 0 }	
	};
	
	while (1) {
		opt = getopt_long(argc, argv, "d:s:S:pDrRvhV", longopts, &opt_idx );
		if (opt == -1)
			break;

		switch (opt) {
		case 0:
			/* long option */
			switch (opt_idx) {
			case 0: device = optarg; break;
			case 1: sensitivity = atoi(optarg); break;
			case 2: press_speed = atoi(optarg); break;
			case 3: show_help = 1; break;
			case 4: press_to_select = 1; break;
			case 5: dragging = 1; break;
			case 6: release_to_select = 1; break;
			case 7: press_right = 1; break;
			case 8: show_version = 1; break;
			}
			break;

		case 'd':
			if (optarg)
				device = optarg;
			break;

		case 's':
			if (optarg)
				sensitivity = atoi(optarg);
			break;

		case 'S':
			if (optarg)
				press_speed = atoi(optarg);
			break;

		case 'p':
			press_to_select = 1;
			break;

		case 'D':
			dragging = 1;
			break;

		case 'r':
			release_to_select = 1;
			break;

		case 'R':
			press_right = 1;
			break;

		case 'v':
			verbosity++;
			break;

		case 'h':
			show_help = 1;
			break;

		case 'V':
			show_version = 1;
			break;

		default:
			break;


		}
	}

	if (show_version) {
		fprintf(stderr, "%s version " STR(VERSION) "\n", basename(argv[0]));
		exit(0);
	}

	if (show_help) {
		fprintf(stderr, "Usage: %s [OPTIONS]\n\n", basename(argv[0]));
		fprintf(stderr, "  -h, --help               show this\n");
		fprintf(stderr, "  -V, --version            show version\n");
		fprintf(stderr, "  -d, --device=?           set device. Autodiscovered if unset\n");
		fprintf(stderr, "  -v                       increase verbosity\n");
		fprintf(stderr, "  -s, --sensitivity=?      set sensitivity, range 1-255\n");
		fprintf(stderr, "  -S, --press-speed=?      set press-speed, range 1-255\n");
		fprintf(stderr, "  -p, --press-to-select    enable press-to-select\n");
		fprintf(stderr, "  -R, --press-right        generate right click instead of left\n");
		fprintf(stderr, "  -D, --dragging           enable dragging\n");
		fprintf(stderr, "  -r, --release-to-select  enable release-to-select\n");
		fprintf(stderr, "\n");
		exit(0);
	}

	if (sensitivity < 1 || sensitivity > 255) {
		fprintf(stderr, "ERROR: sensitivity not within valid range 1-255\n");
		exit(1);
	}

	if (press_speed < 1 || press_speed > 255) {
		fprintf(stderr, "ERROR: press-speed not within valid range 1-255\n");
		exit(1);
	}

	if (device) {
		fd = open(device, O_RDWR|O_NONBLOCK);
	} else {
		fd = find_device();
	}

	if (fd<=0) {
		fprintf(stderr, "No device found\n");
		exit(1);
	}

	buf[0] = 4;
	buf[1]  = press_to_select   ? 0x01 : 0x02;
	buf[1] |= dragging          ? 0x04 : 0x08;
	buf[1] |= release_to_select ? 0x10 : 0x20;
	buf[1] |= press_right       ? 0x80 : 0x40;
	buf[2] = 0x03;
	buf[3] = sensitivity;
	buf[4] = press_speed;

	if (verbosity > 0) {
#define bool2str(x) ((x)?"True":"False")
		fprintf(stderr, "Setting values:\n");
		fprintf(stderr, "-----------------------------\n");
		fprintf(stderr, "sensitivity:            %d\n", sensitivity);
		fprintf(stderr, "press-to-select:        %s\n", bool2str(press_to_select));
		fprintf(stderr, "    press-speed:        %d\n", press_speed);
		fprintf(stderr, "    press-right:        %s\n", bool2str(press_right));
		fprintf(stderr, "    dragging:           %s\n", bool2str(dragging));
		fprintf(stderr, "    release-to-select:  %s\n", bool2str(release_to_select));
		fprintf(stderr, "-----------------------------\n");
#undef bool2str
	}

	ret = ioctl(fd, HIDIOCSFEATURE(5), buf);

	if (ret < 0) {
		close(fd);
		perror("Could not set values");
		//fprintf(stderr, "Error, could not set values (%d)\n", ret);
		exit(1);
	}

	close(fd);
}


tpkbdctl/cmd.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# vim:set sw=4 ts=4 et:

from __future__ import print_function

import sys
import re
from optparse import OptionParser
from tpkbdctl import TpkbdCtl, HidrawDevice

def parse_choice(val):
    v = val[0].lower()
    if v == 'y' or v == 't' or v == '1':
        return True
    return False


def main():
    p = OptionParser(usage='usage: %prog [options]')
    p.add_option('-l', '--list', action='store_true', dest='list',
                help='List all available devices')
    p.add_option('-d', '--device', action='append', type='string', dest='device',
                help='Specify device. Format as printed with --list')
    p.add_option('-s', '--sensitivity', action='store', type='int', dest='sensitivity',
                help='Set trackpoint sensitivity. Range 1-255')
    p.add_option('-S', '--press-speed', action='store', type='int', dest='press_speed', metavar='SPEED',
                help='Set press-speed. Range 1-255')
    p.add_option('-p', '--press-to-select', type='string', action='store', dest='press_to_select', metavar='?',
                 help='Enable press-to-select? (y/n)')
    p.add_option('-R', '--press-right', type='string', action='store', dest='press_right', metavar='?',
                 help='Enable press-right? (y/n)')
    p.add_option('-D', '--dragging', type='string', action='store', dest='dragging', metavar='?',
                 help='Enable dragging? (y/n)')
    p.add_option('-r', '--release-to-select', type='string', action='store', dest='release_to_select', metavar='?',
                 help='Enable release-to-select? (y/n)')


    (options, args) = p.parse_args(sys.argv)


    tpkbdctl = TpkbdCtl()

    if options.device:
        for d in options.device:
            m = re.match('/devices/.*/([0-9a-fA-F]{4}):17EF:6009\.([0-9a-fA-F]{4})', d)
            if m:
                tpkbdctl.probe_device('%04X:17EF:6009.%04X' % (int(m.group(1), 16), int(m.group(2), 16)))
                continue

            m = re.match('tpkbd:([0-9a-fA-F]+):([0-9a-fA-F]+)', d)
            if m:
                tpkbdctl.probe_device('%04X:17EF:6009.%04X' % (int(m.group(1), 16), int(m.group(2), 16)))
                continue

            m = re.match('hidraw:(.*)', d)
            if m:
                tpkbdctl.devices.append(HidrawDevice('/dev/hidraw{0:s}'.format(m.group(1)))) # no validation on purpose
                continue

            m = re.match('/dev/hidraw([0-9]+)', d)
            if m:
                tpkbdctl.devices.append(HidrawDevice('/dev/hidraw{0:s}'.format(m.group(1)))) # no validation on purpose
                continue

    else:
        tpkbdctl.find_devices()

    if options.list:
        for d in tpkbdctl.devices:
            print(str(d))
        exit(0)

    if options.sensitivity != None:
        s = options.sensitivity
        if 0 < s <= 255:
            for d in tpkbdctl.devices: d.sensitivity = s
        else:
            print('Sensitivity value out of range, ignored', file=sys.stderr)

    if options.press_speed != None:
        s = options.press_speed
        if 0 < s <= 255:
            for d in tpkbdctl.devices: d.press_speed = s
        else:
            print('Press-speed value out of range, ignored', file=sys.stderr)

    if options.press_to_select:
        v = parse_choice(options.press_to_select)
        for d in tpkbdctl.devices: d.press_to_select = v

    if options.press_right:
        v = parse_choice(options.press_right)
        for d in tpkbdctl.devices: d.press_right = v

    if options.dragging:
        v = parse_choice(options.dragging)
        for d in tpkbdctl.devices: d.dragging = v

    if options.release_to_select:
        v = parse_choice(options.release_to_select)
        for d in tpkbdctl.devices: d.release_to_select = v



tpkbdctl/__init__.py
# -*- coding: utf-8 -*-
# vim:set sw=4 ts=4 et:

from os import listdir, sep, access, W_OK
from os.path import isdir, isfile, islink, realpath, normpath, basename
from os.path import join as join_path
from struct import pack
from fcntl import ioctl
import re


class HidrawDevice(object):

    can_get = False

    def __init__(self, hidraw_dev):
        #if not access(hidraw_dev, W_OK):
        #    raise RuntimeError('No write access. Maybe run as root?')
        self.hidraw_dev = hidraw_dev
        self._sensitivity = 180
        self._press_speed = 180
        self._press_to_select = False
        self._press_right = False
        self._dragging = False
        self._release_to_select = False

    def __repr__(self):
        return '<HidrawDevice "%s">' % self.hidraw_dev

    def __str__(self):
        return 'hidraw:%s (write-only)' % self.hidraw_dev

    def _write_settings(self):
        props = 0
        props |= 0x01 if self._press_to_select else 0x02
        props |= 0x04 if self._dragging else 0x08
        props |= 0x10 if self._release_to_select else 0x20
        props |= 0x80 if self._press_right else 0x40
        with open(self.hidraw_dev, 'w') as fd:
            ioctl(fd, 0xc0054806, pack('BBBBB', 4, props, 3, self._sensitivity, self._press_speed))


    def get_attr(self):
        raise RuntimeError('Cannot get, only set')

    def set_sensitivity(self, value):
        self._sensitivity = value
        self._write_settings()

    def set_press_speed(self, value):
        self._press_speed = value
        self._write_settings()

    def set_press_to_select(self, value):
        self._press_to_select = value
        self._write_settings()

    def set_press_right(self, value):
        self._press_right = value
        self._write_settings()

    def set_dragging(self, value):
        self._dragging = value
        self._write_settings()

    def set_release_to_select(self, value):
        self._release_to_select = value
        self._write_settings()

    sensitivity = property(get_attr, set_sensitivity)
    press_speed = property(get_attr, set_press_speed)
    press_to_select = property(get_attr, set_press_to_select)
    press_right = property(get_attr, set_press_right)
    dragging = property(get_attr, set_dragging)
    release_to_select = property(get_attr, set_release_to_select)



class TpkbdDevice(object):

    can_get = True

    def __init__(self, hid_path):
        #if not access(hid_path, W_OK):
        #    raise RuntimeError('No write access. Maybe run as root?')
        self.hid_path = hid_path

    def __repr__(self):
        return '<TpkbdDevice "%s">' % basename(self.hid_path)

    def __str__(self):
        [a,b,c,d] = re.split(':|\.', basename(self.hid_path))
        return 'tpkbd:%x:%x' % (int(a, 16), int(d, 16))

    def get_sensitivity(self):
        return int(open(join_path(self.hid_path, 'sensitivity')).readline())

    def set_sensitivity(self, value):
        with open(join_path(self.hid_path, 'sensitivity'), 'w') as fd:
            fd.write(str(value))

    def get_press_speed(self):
        return int(open(join_path(self.hid_path, 'press_speed')).readline())

    def set_press_speed(self, value):
        with open(join_path(self.hid_path, 'press_speed'), 'w') as fd:
            fd.write(str(value))

    def get_press_to_select(self):
        return int(open(join_path(self.hid_path, 'press_to_select')).readline()) == 1

    def set_press_to_select(self, value):
        with open(join_path(self.hid_path, 'press_to_select'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_press_right(self):
        return int(open(join_path(self.hid_path, 'press_right')).readline()) == 1

    def set_press_right(self, value):
        with open(join_path(self.hid_path, 'press_right'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_dragging(self):
        return int(open(join_path(self.hid_path, 'dragging')).readline()) == 1

    def set_dragging(self, value):
        with open(join_path(self.hid_path, 'dragging'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_release_to_select(self):
        return int(open(join_path(self.hid_path, 'release_to_select')).readline()) == 1

    def set_release_to_select(self, value):
        with open(join_path(self.hid_path, 'release_to_select'), 'w') as fd:
            fd.write('1' if value else '0')

    sensitivity = property(get_sensitivity, set_sensitivity)
    press_speed = property(get_press_speed, set_press_speed)
    press_to_select = property(get_press_to_select, set_press_to_select)
    press_right = property(get_press_right, set_press_right)
    dragging = property(get_dragging, set_dragging)
    release_to_select = property(get_release_to_select, set_release_to_select)



class TpkbdCtl(object):

    __hid_path__ = '/sys/bus/hid/devices'
    __dev_path__ = '/dev'


    def __init__(self):
        self._check_prereqs()
        self.devices = []


    def _check_prereqs(self):
        pass


    def find_devices(self):
        self.devices = []
        devs = listdir(self.__hid_path__)
        for dev in devs:
             self.probe_device(dev)


    def _check_interface(self, dev):
        while not isfile(join_path(dev, 'bInterfaceClass')):
            if dev == '/sys/devices' or len(dev) < 2:
                raise RuntimeError('USB device info not found in sysfs')
            dev = realpath(join_path(dev, '..'))

        
        intclass = open(join_path(dev, 'bInterfaceClass')).readline()
        if not re.match(r'^03\s*$', intclass):
            return False

        intnum = open(join_path(dev, 'bInterfaceNumber')).readline()
        if not re.match(r'^01\s*$', intnum):
            return False

        return True


    def probe_device(self, dev):
        if not re.match(r'^....:17EF:6009\.....$', dev):
            return False

        hid_path = realpath(join_path(self.__hid_path__, dev))
        hidraw_path = join_path(hid_path, 'hidraw')
        if not isdir(hidraw_path):
            return False

        hidraw_name = listdir(hidraw_path)[0]
        if not isdir(join_path(hidraw_path, hidraw_name)):
            return False

        if not self._check_interface(hid_path):
            return False

        if isfile(join_path(hid_path, 'sensitivity')):
            self.devices.append(TpkbdDevice(hid_path))
        else:
            hidraw_dev = join_path(self.__dev_path__, hidraw_name)
            self.devices.append(HidrawDevice(hidraw_dev))
        return True


    def __repr__(self):
        return '<Tpkbdctl>'




tpkbdctl.py
#!/usr/bin/python

# -*- coding: utf-8 -*-
# vim:set sw=4 ts=4 et:

from os import listdir, sep, access, W_OK
from os.path import isdir, isfile, islink, realpath, normpath, basename
from os.path import join as join_path
from struct import pack
from fcntl import ioctl
from optparse import OptionParser
import sys
import re


class HidrawDevice(object):

    can_get = False

    def __init__(self, hidraw_dev):
        #if not access(hidraw_dev, W_OK):
        #    raise RuntimeError('No write access. Maybe run as root?')
        self.hidraw_dev = hidraw_dev
        self._sensitivity = 180
        self._press_speed = 180
        self._press_to_select = False
        self._press_right = False
        self._dragging = False
        self._release_to_select = False

    def __repr__(self):
        return '<HidrawDevice "%s">' % self.hidraw_dev

    def __str__(self):
        return 'hidraw:%s (write-only)' % self.hidraw_dev

    def _write_settings(self):
        props = 0
        props |= 0x01 if self._press_to_select else 0x02
        props |= 0x04 if self._dragging else 0x08
        props |= 0x10 if self._release_to_select else 0x20
        props |= 0x80 if self._press_right else 0x40
        with open(self.hidraw_dev, 'w') as fd:
            ioctl(fd, 0xc0054806, pack('BBBBB', 4, props, 3, self._sensitivity, self._press_speed))


    def get_attr(self):
        raise RuntimeError('Cannot get, only set')

    def set_sensitivity(self, value):
        self._sensitivity = value
        self._write_settings()

    def set_press_speed(self, value):
        self._press_speed = value
        self._write_settings()

    def set_press_to_select(self, value):
        self._press_to_select = value
        self._write_settings()

    def set_press_right(self, value):
        self._press_right = value
        self._write_settings()

    def set_dragging(self, value):
        self._dragging = value
        self._write_settings()

    def set_release_to_select(self, value):
        self._release_to_select = value
        self._write_settings()

    sensitivity = property(get_attr, set_sensitivity)
    press_speed = property(get_attr, set_press_speed)
    press_to_select = property(get_attr, set_press_to_select)
    press_right = property(get_attr, set_press_right)
    dragging = property(get_attr, set_dragging)
    release_to_select = property(get_attr, set_release_to_select)



class TpkbdDevice(object):

    can_get = True

    def __init__(self, hid_path):
        #if not access(hid_path, W_OK):
        #    raise RuntimeError('No write access. Maybe run as root?')
        self.hid_path = hid_path

    def __repr__(self):
        return '<TpkbdDevice "%s">' % basename(self.hid_path)

    def __str__(self):
        [a,b,c,d] = re.split(':|\.', basename(self.hid_path))
        return 'tpkbd:%x:%x' % (int(a, 16), int(d, 16))

    def get_sensitivity(self):
        return int(open(join_path(self.hid_path, 'sensitivity')).readline())

    def set_sensitivity(self, value):
        with open(join_path(self.hid_path, 'sensitivity'), 'w') as fd:
            fd.write(str(value))

    def get_press_speed(self):
        return int(open(join_path(self.hid_path, 'press_speed')).readline())

    def set_press_speed(self, value):
        with open(join_path(self.hid_path, 'press_speed'), 'w') as fd:
            fd.write(str(value))

    def get_press_to_select(self):
        return int(open(join_path(self.hid_path, 'press_to_select')).readline()) == 1

    def set_press_to_select(self, value):
        with open(join_path(self.hid_path, 'press_to_select'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_press_right(self):
        return int(open(join_path(self.hid_path, 'press_right')).readline()) == 1

    def set_press_right(self, value):
        with open(join_path(self.hid_path, 'press_right'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_dragging(self):
        return int(open(join_path(self.hid_path, 'dragging')).readline()) == 1

    def set_dragging(self, value):
        with open(join_path(self.hid_path, 'dragging'), 'w') as fd:
            fd.write('1' if value else '0')

    def get_release_to_select(self):
        return int(open(join_path(self.hid_path, 'release_to_select')).readline()) == 1

    def set_release_to_select(self, value):
        with open(join_path(self.hid_path, 'release_to_select'), 'w') as fd:
            fd.write('1' if value else '0')

    sensitivity = property(get_sensitivity, set_sensitivity)
    press_speed = property(get_press_speed, set_press_speed)
    press_to_select = property(get_press_to_select, set_press_to_select)
    press_right = property(get_press_right, set_press_right)
    dragging = property(get_dragging, set_dragging)
    release_to_select = property(get_release_to_select, set_release_to_select)



class TpkbdCtl(object):

    __hid_path__ = '/sys/bus/hid/devices'
    __dev_path__ = '/dev'


    def __init__(self):
        self._check_prereqs()
        self.devices = []


    def _check_prereqs(self):
        pass


    def find_devices(self):
        self.devices = []
        devs = listdir(self.__hid_path__)
        for dev in devs:
            d = self.probe_device(dev)
            if d != None:
                self.devices.append(d)


    def _check_interface(self, dev):
        while not isfile(join_path(dev, 'bInterfaceClass')):
            if dev == '/sys/devices' or len(dev) < 2:
                raise RuntimeError('USB device info not found in sysfs')
            dev = realpath(join_path(dev, '..'))

        
        intclass = open(join_path(dev, 'bInterfaceClass')).readline()
        if not re.match(r'^03$', open(join_path(dev, 'bInterfaceClass')).readline()):
            return False
        if not re.match(r'^01$', open(join_path(dev, 'bInterfaceNumber')).readline()):
            return False

        return True


    def probe_device(self, dev):
        if not re.match(r'^....:17EF:6009\.....$', dev):
            return None

        hid_path = realpath(join_path(self.__hid_path__, dev))
        hidraw_path = join_path(hid_path, 'hidraw')
        if not isdir(hidraw_path):
            return None

        hidraw_name = listdir(hidraw_path)[0]
        if not isdir(join_path(hidraw_path, hidraw_name)):
            return None

        if not self._check_interface(hid_path):
            return None

        #print 'driver: %s' % basename(realpath(join_path(hid_path, 'driver/module')))
        if isfile(join_path(hid_path, 'sensitivity')):
            return TpkbdDevice(hid_path)
        else:
            hidraw_dev = join_path(self.__dev_path__, hidraw_name)
            return HidrawDevice(hidraw_dev)


    def __repr__(self):
        return '<Tpkbdctl>'


def parse_choice(val):
    v = val[0].lower()
    if v == 'y' or v == 't' or v == '1':
        return True
    return False


def tpkbdctl(argv):
    p = OptionParser(usage='usage: %prog [options]')
    p.add_option('-l', '--list', action='store_true', dest='list',
                help='List all available devices')
    p.add_option('-d', '--device', action='append', type='string', dest='device',
                help='Specify device. Format as printed with --list')
    p.add_option('-s', '--sensitivity', action='store', type='int', dest='sensitivity',
                help='Set trackpoint sensitivity. Range 1-255')
    p.add_option('-S', '--press-speed', action='store', type='int', dest='press_speed', metavar='SPEED',
                help='Set press-speed. Range 1-255')
    p.add_option('-p', '--press-to-select', type='string', action='store', dest='press_to_select', metavar='?',
                 help='Enable press-to-select? (y/n)')
    p.add_option('-R', '--press-right', type='string', action='store', dest='press_right', metavar='?',
                 help='Enable press-right? (y/n)')
    p.add_option('-D', '--dragging', type='string', action='store', dest='dragging', metavar='?',
                 help='Enable dragging? (y/n)')
    p.add_option('-r', '--release-to-select', type='string', action='store', dest='release_to_select', metavar='?',
                 help='Enable release-to-select? (y/n)')


    (options, args) = p.parse_args(argv)


    tpkbdctl = TpkbdCtl()

    if options.device:
        for d in options.device:
            m = re.match('tpkbd:([0-9a-fA-F]+):([0-9a-fA-F]+)', d)
            if m:
                tpkbdctl.probe_device('%04X:17EF:6009.%04X' % (int(m.groups(1), 16), int(m.groups(2), 16)))
            
            m = re.match('hidraw:(.*)', d)
            if m:
                tpkbdctl.devices.append(HidrawDevice(str(m.group(1)))) # no validation on purpose
    else:
        tpkbdctl.find_devices()

    if options.list:
        for d in tpkbdctl.devices:
            print str(d)
        exit(0)

    if options.sensitivity != None:
        s = options.sensitivity
        if 0 < s <= 255:
            for d in tpkbdctl.devices: d.sensitivity = s
        else:
            print >>sys.stderr, 'Sensitivity value out of range, ignored'

    if options.press_speed != None:
        s = options.press_speed
        if 0 < s <= 255:
            for d in tpkbdctl.devices: d.press_speed = s
        else:
            print >>sys.stderr, 'Press-speed value out of range, ignored'

    if options.press_to_select:
        v = parse_choice(options.press_to_select)
        for d in tpkbdctl.devices: d.press_to_select = v

    if options.press_right:
        v = parse_choice(options.press_right)
        for d in tpkbdctl.devices: d.press_right = v

    if options.dragging:
        v = parse_choice(options.dragging)
        for d in tpkbdctl.devices: d.dragging = v

    if options.release_to_select:
        v = parse_choice(options.release_to_select)
        for d in tpkbdctl.devices: d.release_to_select = v


if __name__ == '__main__':
    tpkbdctl(sys.argv)
